universal-login-relayer/lib/services/IdentityService.js

```diff
@@ -31,17 +32,26 @@ class IdentityService {
   }
 
   async executeSigned(contractAddress, message) {
-    const contract = new ethers.Contract(contractAddress, this.abi, this.wallet);
-    const addKeySighash = new Interface(Identity.interface).functions.addKey.sighash;
-    if (message.to === contractAddress && message.data.slice(0, addKeySighash.length) === addKeySighash) {
-      const [address] = (this.codec.decode(['bytes32', 'uint256', 'uint256'], message.data.replace(addKeySighash.slice(2), '')));
-      const key = utils.hexlify(utils.stripZeros(address));
-      await this.authorisationService.removeRequest(contractAddress, key);
-      const transaction = await contract.executeSigned(message.to, message.value, message.data, message.nonce, message.gasToken, message.gasPrice, message.gasLimit, message.signature);
-      this.hooks.emit('added', key);
-      return transaction;
+    const {data} = new Interface(Identity.interface).functions.executeSigned(message.to, message.value, message.data, message.nonce, message.gasToken, message.gasPrice, message.gasLimit, message.signature);
+    const transaction = {
+      value: 0,
+      to: contractAddress,
+      data,
+      ...defaultDeployOptions
+    };
+    const estimateGas = await this.wallet.estimateGas(transaction);
+    if (isEnoughGasLimit(estimateGas, message.gasLimit) && await hasEnoughToken(message.gasToken, contractAddress, message.gasLimit, this.provider)) {
+      const addKeySighash = new Interface(Identity.interface).functions.addKey.sighash;
+      if (message.to === contractAddress && message.data.slice(0, addKeySighash.length) === addKeySighash) {
```

Comment:
We should extract method here, i.e: `isAddKeyCall(message.data)`

Reply:
✅